pkg/karmadactl/top/top.go

```diff
@@ -56,3 +64,54 @@ func SupportedMetricsAPIVersionAvailable(discoveredAPIGroups *metav1.APIGroupLis
 	}
 	return false
 }
+
+func GenClusterList(clientSet karmadaclientset.Interface, clusters []string) ([]string, error) {
+	if len(clusters) != 0 {
+		return clusters, nil
+	}
+
+	clusterList, err := clientSet.ClusterV1alpha1().Clusters().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list all member clusters in control plane, err: %w", err)
+	}
+
+	for i := range clusterList.Items {
+		clusters = append(clusters, clusterList.Items[i].Name)
+	}
+	return clusters, nil
+}
+
+func GetCoreV1ClientAndMetricsClient(f util.Factory,
+	cluster string, useProtocolBuffers bool) (v1.CoreV1Interface, metricsclientset.Interface, error) {
+	memberFactory, err := f.FactoryForMemberCluster(cluster)
+	if err != nil {
+		return nil, nil, err
+	}
+	clientset, err := memberFactory.KubernetesClientSet()
+	if err != nil {
+		return nil, nil, err
+	}
+	discoveryClient := clientset.DiscoveryClient
+	apiGroups, err := discoveryClient.ServerGroups()
+	if err != nil {
+		return nil, nil, err
+	}
+	metricsAPIAvailable := SupportedMetricsAPIVersionAvailable(apiGroups)
+	if !metricsAPIAvailable {
+		return nil, nil, fmt.Errorf("Metrics API not available")
+	}
+
+	config, err := memberFactory.ToRESTConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+	if useProtocolBuffers {
+		config.ContentType = "application/vnd.kubernetes.protobuf"
+	}
+	metricsClient, err := metricsclientset.NewForConfig(config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return clientset.CoreV1(), metricsClient, nil
```

Comment:
I rarely encounter this type of return; I prefer to return clientset instead of `clientset.CoreV1()`.

Reply:
Nice suggestion! I prefer it too.